### PR TITLE
Optimize process pull responses

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Large numbers of pull responses are processed individually causing extra locking and low performance.

#### Summary of Changes

Organize pull responses by validator and process in batches.

Fixes #

### CODE REVIEW

**Comment by sakridge** [comment]
gossip-opt

Seems to be working: 10:32 is where I switched to this version. Insert failures are lower by about half.

Processing time is down. Before process_pull_responses is around 30ms. Now, filter is around 5ms and process is around 10ms.

gossip-opt2

File: core/src/crds_gossip_pull.rs [hunk 1]
```diff
@@ -262,42 +267,71 @@ impl CrdsGossipPull {
...
+        crds: &mut Crds,
+        from: &Pubkey,
+        responses: Vec<VersionedCrdsValue>,
+        responses_no_timeout: Vec<VersionedCrdsValue>,
+        now: u64,
+        stats: &mut ProcessPullStats,
+    ) {
+        let mut owners = HashSet::new();
+        for r in responses_no_timeout {
+            stats.failed += crds.insert_versioned(r).is_err() as usize;
```
Comment (carllin) [inline core/src/crds_gossip_pull.rs:313]: Do we want to distinguish between the different types of failures? (timeout vs duplicate)

Comment (sakridge) [reply core/src/crds_gossip_pull.rs:313]: ok

File: core/src/crds_gossip_pull.rs [hunk 2]
```diff
@@ -231,19 +238,17 @@ impl CrdsGossipPull {
...
-        response: Vec<CrdsValue>,
+        responses: Vec<CrdsValue>,
         now: u64,
-    ) -> (usize, usize, usize) {
-        let mut failed = 0;
-        let mut timeout_count = 0;
-        let mut success = 0;
-        for r in response {
+        stats: &mut ProcessPullStats,
+    ) -> (Vec<VersionedCrdsValue>, Vec<VersionedCrdsValue>) {
```
Comment (carllin) [inline core/src/crds_gossip_pull.rs:248]: Could we add a comment about what the return values represent?

File: core/src/crds_gossip_pull.rs [hunk 3]
```diff
@@ -231,19 +238,17 @@ impl CrdsGossipPull {
...
+        let mut versioned = vec![];
+        let mut versioned_no_timestamp = vec![];
```
Comment (carllin) [inline core/src/crds_gossip_pull.rs:250]: is `versioned_no_timestamp` all the expired/invalid timestamp labels? Maybe a better name would be `versioned_expired_timestamp`?

**Comment by sakridge** [comment]
pull_resp_process_time

pull_resp_success

before/after for pull request success/fail and processing time.

**Comment by carllin** [comment]
Do you think the pull response failure rate/processing time (excluding filter) would be even lower if we maintained a `HashMap<Label, Timestamp>` and checked before `would_insert` here: https://github.com/solana-labs/solana/pull/10460/files#diff-8b2c295066af153b1108d9db32997509R284 and here: https://github.com/solana-labs/solana/pull/10460/files#diff-8b2c295066af153b1108d9db32997509R293 to see if a previous instance of the label with a later wallclock has already been added to the `versioned` or `versioned_no_timestamp`. Might help if a lot of those remaining failures are from duplicates within the same batch and are getting dumped b/c of this: https://github.com/solana-labs/solana/pull/10460/files#diff-ab09c2bf25eb0cc66e070cb4fbf14c66R106

**Comment by sakridge** [comment]
> 
> 
> Do you think the pull response failure rate/processing time (excluding filter) would be even lower if we maintained a `HashMap<Label, Timestamp>` and checked before `would_insert` here: https://github.com/solana-labs/solana/pull/10460/files#diff-8b2c295066af153b1108d9db32997509R284 and here: https://github.com/solana-labs/solana/pull/10460/files#diff-8b2c295066af153b1108d9db32997509R293 to see if a previous instance of the label with a later wallclock has already been added to the `versioned` or `versioned_no_timestamp`. Might help if a lot of those remaining failures are from duplicates within the same batch and are getting dumped b/c of this: https://github.com/solana-labs/solana/pull/10460/files#diff-ab09c2bf25eb0cc66e070cb4fbf14c66R106

Potentially, yes. Probably have to write some benchmarks to compare.

Two things I worry about that approach is you have the extra work of lookup/insert into that per-thread hashmap, although it is fully parallel.

Also, the way the threads are doing work now, I'm not sure how many duplicate items there are in a single thread's work batch. This is solvable, but might be some extra work.

**Comment by aeyakovenko** [comment]
can we prioratize the most recent vote?

**Comment by sakridge** [comment]
> can we prioratize the most recent vote?

Yea. I thought of prioritizing just by how recent the responses are, but that could be an extra optimization. I'll experiment with it outside this PR.

**Pull request merged**